File: src/com/colacelli/ircbot/plugins/help/PluginHelper.java
package com.colacelli.ircbot.plugins.help;

import java.util.ArrayList;
import java.util.Comparator;

public class PluginHelper {
    private static PluginHelper instance;
    private ArrayList<PluginHelp> helps;

    private PluginHelper() {
        helps = new ArrayList<>();
    }

    public static PluginHelper getInstance() {
        if (instance == null) instance = new PluginHelper();
        return instance;
    }

    public void addHelp(PluginHelp help) {
        helps.add(help);
    }

    public void removeHelp(String command) {
        for (int i = 0; i < helps.size(); i++) {
            PluginHelp help = helps.get(i);
            if (help.getCommand().toUpperCase().equals(command.toUpperCase())) helps.remove(help);
        }
    }

    public ArrayList<String> getCommands(int access) {
        ArrayList<String> helpTexts = new ArrayList<>();
        this.helps.sort(Comparator.comparing(PluginHelp::getCommand));
        this.helps.forEach(help -> {
            if (help.getAccess() <= access) {
                helpTexts.add(help.getCommand());
            }
        });

        return helpTexts;
    }

    public ArrayList<String> getHelp(int access, String command) {
        ArrayList<String> helpTexts = new ArrayList<>();
        this.helps.sort(Comparator.comparing(PluginHelp::getCommand));
        this.helps.forEach(help -> {
            if (help.getAccess() <= access && (help.getCommand().startsWith(command.toLowerCase()) || help.getCommand().startsWith("." + command.toLowerCase()))) {
                helpTexts.add(help.toString());
            }
        });

        return helpTexts;
    }
}

File: src/com/colacelli/ircbot/plugins/help/PluginHelper.kt
package com.colacelli.ircbot.plugins.help

import com.colacelli.ircbot.plugins.access.IRCBotAccess

class PluginHelper {
    private val helps = ArrayList<PluginHelp>()

    private object Singleton {
        val instance = PluginHelper()
    }

    companion object {
        val instance by lazy {
            Singleton.instance
        }
    }

    fun addHelp(help : PluginHelp) {
        helps.add(help)
        helps.sortBy { it.command }
    }

    fun removeHelp(help : PluginHelp) {
        helps.remove(help)
    }

    fun removeHelp(command : String) {
        helps.removeAll(helps.filter {
            it.command.toLowerCase() == command.toLowerCase()
        })
    }

    fun getCommands(access : IRCBotAccess.Level = IRCBotAccess.Level.USER) : ArrayList<String> {
        val commands = ArrayList<String>()
        helps.forEach {
            // Prevent dupes
            if (commands.indexOf(it.baseCommand) == -1) {
                if (it.access.level < access.level) commands.add(it.baseCommand)
            }
        }

        return commands
    }

    fun getHelp(access : IRCBotAccess.Level = IRCBotAccess.Level.USER, command : String) : ArrayList<String> {
        val texts = ArrayList<String>()
        helps.forEach {
            if (it.command.startsWith(command) && it.access.level <= access.level) texts.add(it.toString())
        }

        return texts
    }
}